### ISSUE TITLE
What is the meaning of perm in ASAPool?

### ISSUE THREAD

**Issue by ZhangGongjie** [rusty1s/pytorch_geometric#2847]
From now on, we recommend using our discussion forum (https://github.com/rusty1s/pytorch_geometric/discussions) for general questions.

## ❓ Questions & Help

What is the meaning of the returning 'perm' in ASAPool? I think it would be much better if such information can in incorporated in docs or annotations in the source codes.

**Comment by rusty1s** [comment]
`perm` denotes the permutation of top-k entries to keep based on the learned scoring function. This information is indeed missing in the documentation, I'm really sorry. I updated the doc, see [here](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#torch_geometric.nn.pool.ASAPooling).

**Comment by ZhangGongjie** [comment]
Thank you so much. Awesome framework! 

May I also ask whether ASAPooling is extremely memory inefficient? Because I apply ASAP with a ratio of 0.5 on nodes of (600, 256), and it gave me OOM error.

**Comment by rusty1s** [comment]
This shouldn't be the case. Can you show me a minimal example to reproduce?

**Comment by ZhangGongjie** [comment]
That is so kind of you. Actually, I am a newbie in graph neural networks. I made a minimal example to reproduce, as follows.

Please download ref_boxes.pt first. (Link: [https://drive.google.com/file/d/18FO-kfKcj1PeJ6QWTyN-bl6YodvJ4kz1/view?usp=sharing])


Codes for reproduction are here:

`
from typing import Union, Optional, Callable

import torch
import torch.nn.functional as F
import torchvision
from torch.nn import Linear
from torch_scatter import scatter

from torch_geometric.nn import LEConv
from torch_geometric.utils import softmax
from torch_geometric.nn.pool.topk_pool import topk
from torch_geometric.utils import add_remaining_self_loops


def box_cxcywh_to_xyxy(x):
    x_c, y_c, w, h = x.unbind(-1)
    b = [(x_c - 0.5 * w), (y_c - 0.5 * h),
         (x_c + 0.5 * w), (y_c + 0.5 * h)]
    return torch.stack(b, dim=-1)


@torch.no_grad()
def convert_object_queries_to_graph(object_query, reference_box):
    device = object_query.device
    batch_size = object_query.shape[0]
    num_node = object_query.shape[1]
    feat_dim = object_query.shape[2]
    assert reference_box.shape[0] == batch_size
    assert reference_box.shape[1] == num_node

    reference_box_iou = [torchvision.ops.box_iou(reference_box[i], reference_box[i]) for i in range(batch_size)]
    reference_box_iou = [(1 - torch.eye(num_node, device=device)) * matrix for matrix in reference_box_iou]  # remove ones on the diagonal

    edge_index = list()
    edge_weight = list()
    for b in range(batch_size):
        index = torch.nonzero(reference_box_iou[b], as_tuple=True)
        weight = reference_box_iou[b][index]
        edge_weight.append(weight)
        edge_index.append(
            torch.stack(index, dim=0) + (b * num_node)
        )

    node_batch = torch.zeros(batch_size * num_node, dtype=torch.long, device=device)
    for b in range(batch_size):
        node_batch[b * num_node: (b + 1) * num_node] = b
    edge_index = torch.cat(edge_index, dim=1)
    edge_weight = torch.cat(edge_weight, dim=0)

    return node_batch, edge_index, edge_weight


class ASAPooling(torch.nn.Module):
    r"""The Adaptive Structure Aware Pooling operator from the
    `"ASAP: Adaptive Structure Aware Pooling for Learning Hierarchical
    Graph Representations" <https://arxiv.org/abs/1911.07979>`_ paper.

    Args:
        in_channels (int): Size of each input sample.
        ratio (float or int): Graph pooling ratio, which is used to compute
            :math:`k = \lceil \mathrm{ratio} \cdot N \rceil`, or the value
            of :math:`k` itself, depending on whether the type of :obj:`ratio`
            is :obj:`float` or :obj:`int`. (default: :obj:`0.5`)
        GNN (torch.nn.Module, optional): A graph neural network layer for
            using intra-cluster properties.
            Especially helpful for graphs with higher degree of neighborhood
            (one of :class:`torch_geometric.nn.conv.GraphConv`,
            :class:`torch_geometric.nn.conv.GCNConv` or
            any GNN which supports the :obj:`edge_weight` parameter).
            (default: :obj:`None`)
        dropout (float, optional): Dropout probability of the normalized
            attention coefficients which exposes each node to a stochastically
            sampled neighborhood during training. (default: :obj:`0`)
        negative_slope (float, optional): LeakyReLU angle of the negative
            slope. (default: :obj:`0.2`)
        add_self_loops (bool, optional): If set to :obj:`True`, will add self
            loops to the new graph connectivity. (default: :obj:`False`)
        **kwargs (optional): Additional parameters for initializing the
            graph neural network layer.
    """
    def __init__(self, in_channels: int, ratio: Union[float, int] = 0.5,
                 GNN: Optional[Callable] = None, dropout: float = 0.0,
                 negative_slope: float = 0.2, add_self_loops: bool = False,
                 **kwargs):
        super(ASAPooling, self).__init__()

        self.in_channels = in_channels
        self.ratio = ratio
        self.negative_slope = negative_slope
        self.dropout = dropout
        self.GNN = GNN
        self.add_self_loops = add_self_loops

        self.lin = Linear(in_channels, in_channels)
        self.att = Linear(2 * in_channels, 1)
        self.gnn_score = LEConv(self.in_channels, 1)
        if self.GNN is not None:
            self.gnn_intra_cluster = GNN(self.in_channels, self.in_channels,
                                         **kwargs)
        self.reset_parameters()

    def reset_parameters(self):
        self.lin.reset_parameters()
        self.att.reset_parameters()
        self.gnn_score.reset_parameters()
        if self.GNN is not None:
            self.gnn_intra_cluster.reset_parameters()

    def forward(self, x, edge_index, edge_weight=None, batch=None):
        N = x.size(0)

        edge_index, edge_weight = add_remaining_self_loops(
            edge_index, edge_weight, fill_value=1, num_nodes=N)

        if batch is None:
            batch = edge_index.new_zeros(x.size(0))

        x = x.unsqueeze(-1) if x.dim() == 1 else x

        x_pool = x
        if self.GNN is not None:
            x_pool = self.gnn_intra_cluster(x=x, edge_index=edge_index,
                                            edge_weight=edge_weight)

        x_pool_j = x_pool[edge_index[0]]
        x_q = scatter(x_pool_j, edge_index[1], dim=0, reduce='max')
        x_q = self.lin(x_q)[edge_index[1]]

        score = self.att(torch.cat([x_q, x_pool_j], dim=-1)).view(-1)
        score = F.leaky_relu(score, self.negative_slope)
        score = softmax(score, edge_index[1], num_nodes=N)

        # Sample attention coefficients stochastically.
        score = F.dropout(score, p=self.dropout, training=self.training)

        v_j = x[edge_index[0]] * score.view(-1, 1)
        x = scatter(v_j, edge_index[1], dim=0, reduce='add')

        # Cluster selection.
        fitness = self.gnn_score(x, edge_index).sigmoid().view(-1)
        perm = topk(fitness, self.ratio, batch)
        x = x[perm] * fitness[perm].view(-1, 1)
        batch = batch[perm]

        # # Graph coarsening.
        # row, col = edge_index
        # A = SparseTensor(row=row, col=col, value=edge_weight,
        #                  sparse_sizes=(N, N))
        # S = SparseTensor(row=row, col=col, value=score, sparse_sizes=(N, N))
        # S = S[:, perm]
        #
        # A = S.t() @ A @ S
        #
        # if self.add_self_loops:
        #     A = A.fill_diag(1.)
        # else:
        #     A = A.remove_diag()
        #
        # row, col, edge_weight = A.coo()
        # edge_index = torch.stack([row, col], dim=0)

        return x, None, None, batch, perm

    def __repr__(self):
        return '{}({}, ratio={})'.format(self.__class__.__name__, self.in_channels, self.ratio)


batch_size = 2
num_node = 300
feat_dim = 256


feature = torch.rand(batch_size, num_node, feat_dim)
boxes = torch.load('ref_boxes.pt')   # (batch_size, num_node, 4) in the format of (x1 y1 x2 y2)
feature = feature.cuda()
boxes = boxes.cuda()
asap = ASAPooling(feat_dim, ratio=0.5)
asap.cuda()

node_batch, edge_index, edge_weight = \
            convert_object_queries_to_graph(feature.detach(), boxes.detach())

feature = feature.reshape(batch_size * num_node, feat_dim)
feature = asap(feature, edge_index, edge_weight, node_batch)

print(feature.shape)

`

Running this code should reproduce the large memory consumption issue. Note that I only include feed-forward operation, and no backward propagation is computed. Solely this pooling operation takes ~1G memory in my GPU.

Thank you so much.

**Comment by rusty1s** [comment]
I cannot access your code snippet.

**Comment by ZhangGongjie** [comment]
I am very sorry. Would you please check the following link?

https://1drv.ms/u/s!AufdUnCZDcQBgplN16BqUwlM8wegBg?e=zn9KTn

**Comment by rusty1s** [comment]
Thanks for the scripts. You are right that `ASAPooling` will consume a lot of memory, especially because your graph is really dense. In particular, there exists about 100K edges in two graphs with node size 300 each. `ASAPooling` will consume attention scores based on an edge representation of shape `[100K, 512]`, which might well explain the high memory consumption.

**Comment by ZhangGongjie** [comment]
Thanks for the explanation. That's really helpful!

I could try to make the graph sparser to ~20 edges per node for a graph of 300 nodes. Do you think this is sparse enough for GNNs to work? Considering that Transformers are intricately densely connected GNNs, do you think GNNs can have any advantages compared with Transformers?

Thank you!

**Comment by rusty1s** [comment]
The sparsity of your graph will definitely help in obtaining a lower memory footprint. GNNs should be advantageous to Transformers in case when edges have a "real" meaning, e.g., there is a clear semantic differences between connected and disconnected nodes.

**Comment by ZhangGongjie** [comment]
That's really insightful! May I respectfully ask do you think my scenario suits this standard, ie, "edges have a "real" meaning, e.g., there is a clear semantic difference between connected and disconnected nodes"?

In my setting, there are multiple boxes referring different locations in an image, and if their IoU > threshold (could be 0, but that would be too dense), there is an edge between two boxes.
